spec/fixtures/labors_model_attributes.csv

```diff
@@ -0,0 +1,33 @@
+project_name,model_name,attribute_name,type,description,restricted,validation
+labors,aspect,monster,parent,,,
+labors,aspect,name,string,,,
+labors,aspect,source,string,,,
+labors,aspect,value,string,,,
+labors,codex,project,parent,,,
+labors,codex,monster,string,,,
+labors,codex,aspect,string,,,
+labors,codex,tome,string,,,
+labors,codex,lore,match,,,
+labors,labor,project,parent,,,
+labors,labor,name,identifier,,,
+labors,labor,monster,child,,,
+labors,labor,number,integer,,,
+labors,labor,completed,boolean,,,
+labors,labor,year,date_time,,,
+labors,labor,prize,table,,,
+labors,labor,contributions,matrix,,,"{""type"":""Array"", ""value"":[""Athens"",""Sparta"",""Sidon"",""Thebes""]}"
```

Comment:
Postgres uses `"` as its escape string for CSVs so we end up with lots of `""` in the validation JSON.

Reply:
What motivated the choice of CSV instead of JSON or YAML? the latter might be less compact in terms of number of lines but I think they would be more legible/editable and the CSV is sparsely populated so a YAML might not be so noisy.

Reply:
I went with CSV because it feeds straight into Postgres' `COPY` command. This lets us take advantage of Postgres' features for importing data, which means less code that we have to maintain in the future. The `COPY` command is also pretty fast so we shouldn't have to worry about the import slowing down the overall test run time.

That being said, maintaining the CSV could also be challenging because it is less legible. Lemme play around with a YAML file and we can compare the two.

Reply:
Ok, I switched it to YAML in https://github.com/mountetna/magma/pull/148/commits/3b5cc2a905de003ce6b6213074fdcce0d8b21da4. If we don't like it, I can undo the commit.

The YAML file is more human readable, but we have to do more work parsing it on import.

https://github.com/mountetna/magma/blob/3b5cc2a905de003ce6b6213074fdcce0d8b21da4/spec/spec_helper.rb#L35-L45

vs.

https://github.com/mountetna/magma/blob/a034bbfddc87fa1edeb6ad745f8fc8ac11ff28cd/spec/spec_helper.rb#L35-L40

It's a bummer we can't leverage Postgres' `COPY` command with the YAML file, but if the YAML is easier to read/edit it might be a worthwhile trade off. What do you think?

Reply:
If it helps there is `Sequel::Model.multi_insert` which you can feed a bunch of hashes & bulk insert (the hashes all have to have the same keys). This uses Postgres' `INSERT`; `COPY` is more efficient, but in this case it is probably not that significant a difference - though we might want to benchmark the difference with ~10k items at some point.

You could also continue to use `COPY` by packing the YAML data into a CSV using `CSV.generate`.

Reply:
Yeah, we're not inserting enough stuff for `COPY` to make a big difference.

Maybe we stick with what we have now. Seems like the best compromise 😄